Start shipping our own type stubs

Typeshed houses very rudimentary type stubs for urllib3 but they're hidden within `requests.packages.urllib3` so Mypy can't take advantage of them when using urllib3 on its own. We could also probably improve on what is already in typeshed by adding better types than `Any` everywhere.

Cool, sounds like a good starting point. Two questions, stemming from my lack of familiarity with stubs and/or `typeshed`.

1. How do you run a type check with the existing `typeshed` [stubs](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3) of `urllib3` since it's in `requests`? Detailed run below, importing `requests` doesn't seem to help.

```
$ cat run.py

import requests
import urllib3

def pool():
    return urllib3.PoolManager()

$ mypy run.py
run.py:3: error: Skipping analyzing 'urllib3': found module but no type hints or library stubs
run.py:3: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 1 source file)
```

2. I presume all the changes would be in the `typeshed` [repo](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3) alone. What's the anticipated flow to check these changes are correct (since `typeshed` normally comes with `mypy`), i.e. clone `typeshed`, move to `MYPYPATH`, run `mypy` on `urrlib3` tests (since stubs won't check the code [itself](https://github.com/python/mypy/issues/5028))?

--- a/noxfile.py
+++ b/noxfile.py
@@ -4,41 +4,6 @@ import subprocess
 
 import nox
 
-# Whenever type-hints are completed on a file it should be added here so that
-# this file will continue to be checked by mypy. Errors from other files are
-# ignored.
-TYPED_FILES = {
-    "src/urllib3/contrib/__init__.py",
-    "src/urllib3/contrib/_securetransport/bindings.py",
-    "src/urllib3/contrib/_securetransport/low_level.py",
-    "src/urllib3/contrib/ntlmpool.py",
-    "src/urllib3/contrib/pyopenssl.py",
-    "src/urllib3/contrib/securetransport.py",
-    "src/urllib3/contrib/socks.py",
-    "src/urllib3/__init__.py",
-    "src/urllib3/connection.py",
-    "src/urllib3/connectionpool.py",
-    "src/urllib3/exceptions.py",
-    "src/urllib3/_collections.py",
-    "src/urllib3/fields.py",
-    "src/urllib3/filepost.py",
-    "src/urllib3/poolmanager.py",
-    "src/urllib3/request.py",
-    "src/urllib3/response.py",
-    "src/urllib3/util/connection.py",
-    "src/urllib3/util/proxy.py",
-    "src/urllib3/util/queue.py",
-    "src/urllib3/util/response.py",
-    "src/urllib3/util/ssl_.py",
-    "src/urllib3/util/ssl_match_hostname.py",
-    "src/urllib3/util/ssltransport.py",
-    "src/urllib3/util/url.py",
-    "src/urllib3/util/request.py",
-    "src/urllib3/util/retry.py",
-    "src/urllib3/util/timeout.py",
-    "src/urllib3/util/util.py",
-    "src/urllib3/util/wait.py",
-}
 SOURCE_FILES = [
     "docs/",
     "dummyserver/",
@@ -146,27 +111,7 @@ def mypy(session):
     session.install("mypy==0.812")
     session.install("idna>=2.0.0")
     session.run("mypy", "--version")
-
-    session.log("mypy --strict src/urllib3")
-    all_errors, errors = [], []
-    process = subprocess.run(
-        ["mypy", "--strict", "src/urllib3"],
-        env=session.env,
-        text=True,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-    )
-    # Ensure that mypy itself ran successfully
-    assert process.returncode in (0, 1)
-
-    for line in process.stdout.split("\n"):
-        all_errors.append(line)
-        filepath = line.partition(":")[0]
-        if filepath.replace(".pyi", ".py") in TYPED_FILES:
-            errors.append(line)
-    session.log(f"all errors count: {len(all_errors)}")
-    if errors:
-        session.error("\n" + "\n".join(sorted(set(errors))))
+    session.run("mypy", "--strict", "src/urllib3")
 
 
 @nox.session

